20231212-modular-models.md

```diff
@@ -207,6 +207,41 @@ type page
 
 Module information will be stored in the JSON representation of the model as metadata. Using that information it should be possible to reconstruct each module, and use module information for authorization decision in the future (e.g. certain applications can only write tuples to certain modules).
 
+This metadata will be stored in the below structure, and will be stored as part of the existing `metadata` for types and relations, and a new `metadata` property will be added for conditions. File and module metadata will only be added to relations that originate from extended types and not to relations that come from the original type.
+
+```json
+{
+  ...,
+  "file": string,
+  "module": string
+}
+```
+
+For example the `organization` type from `core.module.fga` would look like below (fields other than `metadata` removed for brevity):
+
+```json
+{
+  "type": "organization",
+  "relations": { ... },
+  "metadata": {
+    "relations": {
+      "member": { ... },
+      "admin": { ... },
+      "can_create_space": {
+        "module": "confluence",
+        "file": "confluence.module.fga"
+      },
+      "can_create_project": {
+        "module": "jira",
+        "file": "jira.module.fga"
+      }
+    },
+    "module": "core",
+    "file": "core.module.fga"
```

Comment:
I propose we encapsulate source information in a nested structure, thus allowing the structure and extension of such information to grow in a scoped manner instead of littering the top-level definition. It also makes the information more strongly typed and consumable for consuming application code.

May I propose:
```json
"metadata": {
  "relations": {...}
},
"source_info": {
    "module": "core",
    "file": "core.module.fga"
}
```

In the future we're have more, such as:
```json
"source_info": {
    "module": "core",
    "file": "core.module.fga",
    "line": {
        "column": 17,
        "row": 2
    }
}
```

The protobuf we introduce for this in our API would match something of the form:
```protobuf
message SourceInformation {
   string module = 1;
   string file = 2;
   SourceLineInfo line = 3; // in the future
}
```

Reply:
I agree on moving the filename into a source information field, but I think `module` should still be classified as metadata due to the potential future importance and requirement for features such as FGA-on-FGA where `module` would become a native/first-class piece of metadata akin to the existing relations metadata.

cc @rhamzeh who can probably articulate the reasoning better than me.

Reply:
+1 to what @ewanharris said.

We can do:
```json
"metadata": {
    "relations": { ... }, // existing field (only relevant for types)
    "module": "core", // added
    "source_info": { // added
        "file": "core.module.fga"
    }
}
```

Let's make sure this is the same structure for types and conditions. relations would follow something similar, but because it would be already nested in the types metadata, there's no need to say that again.

So for relations it would look like:
```json
"admin": {
    "directly_related_user_types": [{ "type": "user" }], // existing field
    "module": "core", // added
    "source_info": { // added
        "file": "core.module.fga"
    }
}
```

Reply:
I've updated the [API PR ](https://github.com/openfga/api/pull/141) and this PR with this new proposal

Reply:
👍 looks great. Thanks for the consideration folks.